### PR TITLE
Nvidia port

### PR DESCRIPTION
**Pull request by amarathe84**

### CODE REVIEW

**Comment by slabasan** [comment]
Related to #19

File: src/variorum/Nvidia/power_features.c [hunk 1]
```diff
@@ -0,0 +1,241 @@
...
+    for (device_index = 0; device_index < 2; device_index++)
+    {
+        for (d = device_index * gpus_per_socket; d < (device_index + 1) * gpus_per_socket; ++d)
+        {
+            nvmlDeviceGetUtilizationRates(m_unit_devices_file_desc[d], &util);
+
+            if (verbose)
+            {
+                fprintf(writedest, "_GPU_UTILIZATION Host: %s, Socket: %d, Device ID: %d, GPU Utilization (%): SM: %d, Memory: %d\n",
+                    hostname, device_index, d, util.gpu, util.memory);
```
Comment (slabasan) [inline src/variorum/Nvidia/power_features.c]: Indent `hostname, device_index, d, util.gpu, util.memory);` on line 227 so it's even with `writedest`.

File: src/variorum/Nvidia/Volta.h [hunk 2]
```diff
@@ -0,0 +1,10 @@
+#ifndef VOLTA_H_INCLUDE
+#define VOLTA_H_INCLUDE
+
+int volta_get_power(int long_ver);
+int volta_get_thermals(int long_ver);
+int volta_get_clocks(int long_ver);
+int volta_get_power_limits(int long_ver);
+int volta_get_gpu_utilization(int long_ver);
+
+#endif
```
Comment (slabasan) [inline src/variorum/Nvidia/Volta.h]: Add extra space between each function declaration in header files:
```
int volta_get_power(int long_ver);

int volta_get_thermals(int long_ver);

int volta_get_clocks(int long_ver);

int volta_get_power_limits(int long_ver);

int volta_get_gpu_utilization(int long_ver);
```

File: src/variorum/Nvidia/power_features.h [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+
+void shutdownNVML(void);
+
+void dump_power_data(FILE *writedest, int verbose);
+void dump_thermal_data(FILE *writedest, int verbose);
+void dump_power_limits(FILE *writedest, int verbose);
+void dump_clocks_data(FILE *writedest, int verbose);
+void dump_gpu_utilization(FILE *writedest, int verbose);
+
+#endif
```
Comment (slabasan) [inline src/variorum/Nvidia/power_features.h]: Add extra line between function declarations as you have done with `initNVML()` and `shutdownNVML()`.

File: src/variorum/variorum.c [hunk 4]
```diff
@@ -611,6 +611,60 @@ int variorum_print_turbo(void)
...
+    {
+        return -1;
+    }
+    err = variorum_exit(__FILE__, __FUNCTION__, __LINE__);
+    if (err)
+    {
+        return -1;
+    }
+    return err;
+
```
Comment (slabasan) [inline src/variorum/variorum.c]: Remove extra new line here.

File: src/variorum/variorum.c [hunk 5]
```diff
@@ -611,6 +611,60 @@ int variorum_print_turbo(void)
...
```
Comment (slabasan) [inline src/variorum/variorum.c]: Remove extra new line here as well.

File: src/variorum/variorum_config.h.in [hunk 6]
```diff
@@ -6,11 +6,11 @@
...
-#cmakedefine VARIORUM_WITH_IBM   @VARIORUM_WITH_IBM@
-#cmakedefine VARIORUM_WITH_GPU   @VARIORUM_WITH_GPU@
-#cmakedefine VARIORUM_DEBUG      @VARIORUM_DEBUG@
-#cmakedefine VARIORUM_LOG        @VARIORUM_LOG@
+#cmakedefine VARIORUM_WITH_INTEL    @VARIORUM_WITH_INTEL@
+#cmakedefine VARIORUM_WITH_AMD      @VARIORUM_WITH_AMD@
+#cmakedefine VARIORUM_WITH_IBM      @VARIORUM_WITH_IBM@
+#cmakedefine VARIORUM_WITH_NVIDIA   @VARIORUM_WITH_NVIDIA@
+#cmakedefine VARIORUM_DEBUG         @VARIORUM_DEBUG@
+#cmakedefine VARIORUM_LOG           @VARIORUM_LOG@
```
Comment (slabasan) [inline src/variorum/variorum_config.h.in]: Can we align the last column with `@VARIORUM_WITH_NVIDIA@`?

File: src/variorum/config_architecture.c [hunk 7]
```diff
@@ -108,6 +115,7 @@ int variorum_detect_arch(void)
     if (g_platform.intel_arch == NULL &&
         g_platform.amd_arch   == NULL &&
         g_platform.ibm_arch   == NULL &&
+        g_platform.nvidia_arch   == NULL &&
```
Comment (slabasan) [inline src/variorum/config_architecture.c]: Can we align all items in this if block (`== NULL`) with `g_platform.nvidia_arch   == NULL &&`?

File: src/variorum/config_architecture.h [hunk 8]
```diff
@@ -66,6 +66,20 @@ enum ibm_arch_e
...
+    VOLTA = 1, // Volta
+};
+
+/// @brief Mapping topology information.
+struct map
+{
+    /// @brief Unique ID of each physical core.
+    int physical_core_idx;
+};
+
```
Comment (slabasan) [inline src/variorum/config_architecture.h]: Do you still need this struct after the recent hwloc change?

File: src/variorum/config_architecture.c [hunk 9]
```diff
@@ -94,9 +98,12 @@ int variorum_detect_arch(void)
 #ifdef VARIORUM_WITH_IBM
     g_platform.ibm_arch = detect_ibm_arch();
 #endif
-#ifdef VARIORUM_WITH_GPU
+#ifdef VARIORUM_WITH_NVIDIA
```
Comment (slabasan) [inline src/variorum/config_architecture.c]: I think this ifdef block can be deleted (it has `detect_gpu_arch()` commented out in any case).

File: src/variorum/Nvidia/power_features.c [hunk 10]
```diff
@@ -0,0 +1,241 @@
...
+
+    /* get handles to all devices */
+    for (d = 0; d < m_total_unit_devices; ++d)
+    {
+        int power;
+
+        result = nvmlDeviceGetHandleByIndex(d, &m_unit_devices_file_desc[d]);
+        if (result != NVML_SUCCESS)
+        {
+            printf("Error: Could not get GPU device handle.\n");
```
Comment (slabasan) [inline src/variorum/Nvidia/power_features.c]: There is a `variorum_error()` for common error reporting.

File: src/variorum/Nvidia/Volta.c [hunk 11]
```diff
@@ -0,0 +1,52 @@
+#include <stdio.h>
```
Comment (slabasan) [inline src/variorum/Nvidia/Volta.c]: Need to insert the SPDX license header to all your new files.

File: host-configs/lassen-rhel-gcc-gpu.cmake [hunk 12]
```diff
@@ -0,0 +1,26 @@
+# c compiler
+#set(CMAKE_C_COMPILER "/usr/tce/packages/clang/clang-coral-2018.08.08/bin/clang" CACHE PATH "")
+set(CMAKE_C_COMPILER "gcc" CACHE PATH "")
+
+# cpp compiler
+#set(CMAKE_CXX_COMPILER "/usr/tce/packages/clang/clang-coral-2018.08.08/bin/clang++" CACHE PATH "")
```
Comment (slabasan) [inline host-configs/lassen-rhel-gcc-gpu.cmake]: Should we keep the clang compilers in this host config?

**Comment by slabasan** [comment]
Reviewed this PR. Overall looks great! Thanks for contributing the Nvidia port. Most comments are just minor changes.

File: src/variorum/Nvidia/power_features.c (hunk 1 above)
Comment (amarathe84) [reply src/variorum/Nvidia/power_features.c]: Fixed.

File: src/variorum/Nvidia/Volta.h (hunk 2 above)
Comment (amarathe84) [reply src/variorum/Nvidia/Volta.h]: Fixed.

File: src/variorum/Nvidia/power_features.h (hunk 3 above)
Comment (amarathe84) [reply src/variorum/Nvidia/power_features.h]: Added extra lines between function declarations.

File: src/variorum/variorum.c (hunk 4 above)
Comment (amarathe84) [reply src/variorum/variorum.c]: Fixed.

File: src/variorum/variorum.c (hunk 5 above)
Comment (amarathe84) [reply src/variorum/variorum.c]: Fixed.

File: src/variorum/config_architecture.c (hunk 7 above)
Comment (amarathe84) [reply src/variorum/config_architecture.c]: Fixed the alignment.

File: src/variorum/config_architecture.h (hunk 8 above)
Comment (amarathe84) [reply src/variorum/config_architecture.h]: Not anymore. Removed the 'map' structure as the hwloc topology can be used now.

File: src/variorum/config_architecture.c (hunk 9 above)
Comment (amarathe84) [reply src/variorum/config_architecture.c]: Good point -- made the change.

File: host-configs/lassen-rhel-gcc-gpu.cmake (hunk 12 above)
Comment (amarathe84) [reply host-configs/lassen-rhel-gcc-gpu.cmake]: Clang is no longer required. Cleaned up the hostfile.

File: src/variorum/Nvidia/power_features.c (hunk 10 above)
Comment (amarathe84) [reply src/variorum/Nvidia/power_features.c]: Fixed.

**Comment by slabasan** [comment]
+1

**Pull request merged**